### PR TITLE
10 implement flist functional list

### PR DESCRIPTION
**Pull request by EvgeniiaGranina**
Code defines an abstract class FList that represents a functional list with generic types. It includes two inner classes, Empty and Cons, which implement the behavior for empty and non-empty lists. Here is a brief description of each part:

Abstract Class FList:
Defines methods such as head(), tail(), prepend(), append(), isEmpty(), size(), map(), and filter().
Has a static method empty() that returns an empty list.
Inner Class Empty:
Implements an empty list.
Methods like head() and tail() throw UnsupportedOperationException because an empty list has no head or tail.
prepend() and append() return a new Cons list with the added element.
isEmpty() returns true and size() returns 0.
map() returns a new empty list and filter() returns itself.
Inner Class Cons:
Implements a non-empty list.
Has fields for head and tail.
prepend() returns a new Cons list with the added element.
append() returns a new Cons list with the added element at the end.
isEmpty() returns false and size() returns the size of the list.
map() returns a new Cons list with the mapped value.
filter() returns a new Cons list with filtered values.
Test Class FListTest:
Contains tests to verify the functionality of FList.
Tests methods like empty(), prepend(), append(), map(), and filter().

### CODE REVIEW

**Review by kirre02 — APPROVED** [review]
LGTM 👍🏽

**Review by Lhlmlund — APPROVED** [review]
Great implementation I think you have Implementing an immutable data structure with functional programming, I can see the correct API operations with good test coverage that covers the entire class.  I would maybe recommend even more  testing that can be implemented to cover even more edge case circumstances

**Review by kappsegla — CHANGES_REQUESTED** [review]
CodeRabbit seems to have some good suggestions that I think we should implement before merging this code. Check under nitpicks

**Comment by kappsegla** [comment]
@EvgeniiaGranina looks like your commits isn't signed so they won't be able to be merged. You will have to redo them using an interactive rebase and sign the new commits :)

**Comment by Raffi02k** [comment]
Good work, i think your description could be cleaner. You could use chat gpt to make the description much cleaner.

- Cleaning and organizing is a practice, not a project. 
@liamr03

File: src/main/java/org/fungover/breeze/collection/FList.java [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+        /**
+         * Returns a new list with the specified element added at the end.
+         * @param element the element to add
+         * @return a new list with the element added at the end
+         */
+        @Override
+        public FList<T> append(T element) {
+            return new Cons<>(head, tail.append(element));
+        }
```
Comment (kappsegla) [inline src/main/java/org/fungover/breeze/collection/FList.java:225]: What will happen when running this on a long list? How many calls to append will happen? Same amount as it already is in the list? Or just one?

Comment (EvgeniiaGranina) [reply src/main/java/org/fungover/breeze/collection/FList.java:225]: public FList<T> append(T element) {
            // Check if the current list is empty
            if (this == Empty.SHARED_EMPTY)
                // If the list is empty, create a new list with the element as the only item
                return new Cons<T>(element, this);
            else
                // If the list is not empty, recursively append the element to the tail
                return new Cons<>(head, tail.append(element));
        }
Can that solve the problem?

Comment (kappsegla) [reply src/main/java/org/fungover/breeze/collection/FList.java]: I think the biggest problem is the non empty part when using a huge list for example.

// If the list is not empty, **recursively** append the element to the tail
return new Cons<>(head, tail.append(element));

But might not be possible to handle that with the current implementation and keeping it immutable?

Comment (kappsegla) [reply src/main/java/org/fungover/breeze/collection/FList.java]: To get a faster append we could store the list in reversed order and do a prepend instead when adding something last. But then we would have to reverse it when reading which can be optimized using a cache so its only the first time it has a performance hit. :)

Maybe we could document in the documentation for the most common methods in the list what their big O cost is. I think the cost of appending something is O(n) for the append method.

Comment (EvgeniiaGranina) [reply src/main/java/org/fungover/breeze/collection/FList.java]: @kappsegla Can you check if it works properly now?

**Review by MarcGust — APPROVED** [review]
Looks clean, well-documented and functional to me.

**Review by kappsegla — APPROVED** [review]

**Comment by kappsegla** [comment]
@coderabbitai full review

File: src/main/java/org/fungover/breeze/collection/FList.java [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+    }
+
+    /**
+     * CachedFList is a wrapper around FList that caches the reversed version of the list.
+     * This improves the performance of append operations by allowing elements to be prepended
+     * to the reversed list.
+     *
+     * @param <T> the type of elements in this list
+     */
+    public static class CachedFList<T> extends FList<T> {
```
Comment (VictoriaNaenfeldtEklund) [inline src/main/java/org/fungover/breeze/collection/FList.java:372]: Is it nessary for this class to be public or could it maybe be private to prevent being passed a FList.empty()?

**Review by VictoriaNaenfeldtEklund — APPROVED** [review]
This looks great! Good job implementing this FList :)
I had a small comment which is more of a question from my part, not anything I think needs to be changed for approval so this is approved !

File: src/main/java/org/fungover/breeze/collection/FList.java (hunk 2 above)
Comment (EvgeniiaGranina) [reply src/main/java/org/fungover/breeze/collection/FList.java:372]: Yes, private can indeed help prevent it from being misused, such as being passed an FList.empty().
